timemachine/cpp/src/potential.cu

```diff
@@ -1,16 +1,112 @@
-#include <iostream>
 #include <memory>
 
 #include "device_buffer.hpp"
 #include "fixed_point.hpp"
 #include "gpu_utils.cuh"
 #include "potential.hpp"
-#include "surreal.cuh"
+
+#include <chrono>
+#include <iostream>
 
 namespace timemachine {
 
 const int Potential::D = 3;
 
+void Potential::execute_batch_host(
+    const int coord_batches,     // Number of batches of coordinates
+    const int N,                 // Number of atoms
+    const int param_batches,     // Number of batches of parameters
+    const int P,                 // Number of parameters
+    const int lambda_batches,    // Number of lambda values
+    const double *h_x,           // [coord_batches, N, 3]
+    const double *h_p,           // [param_batches, P]
+    const double *h_box,         // [coord_batches, 3, 3]
+    const double *lambdas,       // [lambda_batches]
+    unsigned long long *h_du_dx, // [coord_batches, param_batches, lambda_batches, N, 3]
+    unsigned long long *h_du_dp, // [coord_batches, param_batches, lambda_batches, P]
+    unsigned long long *h_du_dl, // [coord_batches, param_batches, lambda_batches, N]
+    unsigned long long *h_u) {   // [coord_batches, param_batches, lambda_batches, N]
```

Comment:
nice! This looks like it implements the "outer/cartesian product" of the batches. Which I think covers most use cases (Eg. computing a frame at different values of lambda). Maybe later on there will be a use case for the "inner-product" but probably lower priority.

Reply:
Actually we already implicitly inner-product the coords and box - (i.e. they're always tied together, we don't evaluate a single conformation over multiple boxes)